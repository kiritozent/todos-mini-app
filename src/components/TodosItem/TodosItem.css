.TodosItemContainer {
  display: flex;
  flex-direction: row;
  justify-content: center;
  align-items: center;
  width: 100%;
}

.TodosItem {
  cursor: pointer;
  padding: 5.5px 11px;
  margin: 4px 0px;
  border-radius: 4px;
  margin-right: 16px;
  flex: 1;
}

.DeleteIcon {
  color: tomato;
  font-size: 16px;
}

.TodosItem:hover {
  border-width: 1px;
  padding: 4.5px 10px;
  border-style: solid;
  border-color: lightgray;
}
