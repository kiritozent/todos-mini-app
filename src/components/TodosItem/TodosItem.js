/* eslint-disable no-nested-ternary */
/* eslint-disable jsx-a11y/no-static-element-interactions */
/* eslint-disable jsx-a11y/click-events-have-key-events */
import React, { useState, useRef, useEffect, useContext } from 'react';
import PropTypes from 'prop-types';
import { Input, Form, Modal } from 'antd';
import { DeleteOutlined, ExclamationCircleOutlined } from '@ant-design/icons';
import './TodosItem.css';
import { StoreContext } from '../../store';
import Spinner from '../Spinner/Spinner';
import { DEFAULT_REDUCERS } from '../../data/const';

const TodosItem = props => {
  const {
    item,
    patchTodos,
    deleteTodos,
    patchTodosRequest,
    deleteTodosRequest,
  } = props;
  const inputRef = useRef();
  const [title, setTitle] = useState();
  const [onEdit, setOnEdit] = useState(false);
  const [deleteLoading, setDeleteLoading] = useState(false);

  useEffect(() => {
    if (onEdit) {
      inputRef.current.focus();
    } else if (item) {
      setTitle(item.title);
    }
  }, [onEdit]);

  useEffect(() => {
    if (item) {
      setOnEdit(false);
      setTitle(item.title);
    }
  }, [item]);

  useEffect(() => {
    if (!deleteTodos.fetching) {
      setDeleteLoading(false);
    }
  }, [deleteTodos.fetching]);

  function onPatchTodos() {
    const data = {
      ...item,
      title,
    };

    patchTodosRequest(data);
  }

  function onDeleteTodos() {
    Modal.confirm({
      title: 'Are you sure delete this todo?',
      icon: <ExclamationCircleOutlined />,
      okText: 'Yes',
      okType: 'danger',
      cancelText: 'No',
      okButtonProps: {
        loading: deleteLoading,
        onClick: () => {
          setDeleteLoading(true);
          deleteTodosRequest(item);
          Modal.destroyAll();
        },
      },
      onCancel() {
        console.log('Cancel');
      },
    });
  }

  if (onEdit)
    return (
      <Form.Item
        className="TodosFormItem"
        validateStatus="validating"
        hasFeedback={patchTodos.fetching}
      >
        <Input
          ref={inputRef}
          value={title}
          onChange={e => setTitle(e.target.value)}
          onBlur={() => setOnEdit(false)}
          onKeyPress={e => {
            if (e.key === 'Enter') {
              onPatchTodos();
            }
          }}
        />
      </Form.Item>
    );
  return (
    <div className="TodosItemContainer">
      <div
        className="TodosItem"
        onClick={() => {
          setOnEdit(true);
        }}
        style={{
          textDecoration: item && item.completed ? 'line-through' : 'none',
        }}
      >
        {title}
      </div>
      {deleteLoading ? (
        <Spinner size="small" />
      ) : (
        <DeleteOutlined
          className="DeleteIcon"
          role="button"
          onClick={onDeleteTodos}
        />
      )}
    </div>
  );
};

TodosItem.propTypes = {
  item: PropTypes.object,
  patchTodos: PropTypes.object,
  deleteTodos: PropTypes.object,
  patchTodosRequest: PropTypes.func,
  deleteTodosRequest: PropTypes.func,
};

TodosItem.defaultProps = {
  item: null,
  patchTodos: DEFAULT_REDUCERS,
  deleteTodos: DEFAULT_REDUCERS,
  patchTodosRequest: () => {},
  deleteTodosRequest: () => {},
};

export default TodosItem;
