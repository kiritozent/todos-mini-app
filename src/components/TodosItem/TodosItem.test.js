import React from 'react';
import { shallow } from 'enzyme';
import TodosItem from './TodosItem';

describe('TodosItem', () => {
  it('render without crashing passed', () => {
    const testCase = shallow(<TodosItem />);
    expect(testCase).toMatchSnapshot();
  });
});
