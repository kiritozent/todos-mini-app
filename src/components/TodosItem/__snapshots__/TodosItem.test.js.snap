// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TodosItem render without crashing passed 1`] = `
<div
  className="TodosItemContainer"
>
  <div
    className="TodosItem"
    onClick={[Function]}
    style={
      Object {
        "textDecoration": "none",
      }
    }
  />
  <ForwardRef(DeleteOutlined)
    className="DeleteIcon"
    onClick={[Function]}
    role="button"
  />
</div>
`;
